### PR TITLE
API boilerplate

### PR DESCRIPTION
**Pull request by mrjvs**
features:
 - bootstrap function at start of process
 - code structure for running api methods outside of the sandbox
 - code structure for creating sandbox-only methods

### CODE REVIEW

File: src/managers/isolate-manager.ts [hunk 1]
```diff
@@ -15,41 +20,58 @@ export function setLimit(
...
+    (ins) => disposeOfInstance(ins.id),
+    timeout,
+  );
   isolates.push(instance);
   return instance;
 }
 
 function handleCatastrophicError(message: string) {
-  // TODO try to shut down as nicely as possible
+  // TODO cleanup?
```
Comment (weeryan17) [inline src/managers/isolate-manager.ts]: Yes

File: src/models/setup-context.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  const jail = ctx.global;
+  await jail.set('global', jail.derefInto());
+
+  // simple console log for debugging purposes
+  // TODO remove this function from production code
+  await registerFunction(ctx, 'log', (msg: any) => {
+    if (stringSchema.validate(msg).error) return false;
+    log.debug('SANDBOX: ' + msg);
+    return true;
+  });
```
Comment (weeryan17) [inline src/models/setup-context.ts]: No reason we can't have this log in prod. Just make prod not print debug logs. We're going to need to add more stuff here for sending logs to the dashboard or other system, but I think it's fine having debug logging sandbox messages.

File: src/api/entrypoint.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  importedData.forEach((data: Schema) => {
+    schemas[data.action] = data;
+  });
+}
+
+/*
+ ** main entrypoint for sandboxed methods to be called
+ ** all parameters of this method are untrusted
+ */
+export async function callApiMethod(payload: any): Promise<any> {
```
Comment (weeryan17) [inline src/api/entrypoint.ts]: ahhh yes, all 1 parameter is untrusted. Is this going to serialize all data into an object, or is this supposed to accept multiple parameters?

**Review by weeryan17 — CHANGES_REQUESTED** [review]

File: src/models/setup-context.ts (hunk 2 above)
Comment (mrjvs) [reply src/models/setup-context.ts]: dont sure I agree, we'll talk with binary about it

File: src/managers/isolate-manager.ts (hunk 1 above)
Comment (mrjvs) [reply src/managers/isolate-manager.ts]: yes

File: src/api/entrypoint.ts (hunk 3 above)
Comment (mrjvs) [reply src/api/entrypoint.ts]: updated comment

**Review by weeryan17 — APPROVED** [review]

**Review by binaryoverload — APPROVED** [review]

**Pull request merged**